##**BOOK RECOMMENDER**
dataset=['https://www.kaggle.com/datasets/dylanjcastillo/7k-books-with-metadata/data?select=books.csv']

---



In [1]:
import pandas as pd

In [2]:
df=pd.read_csv('/content/books.csv')
df

isbn13      isbn10                      title  \
0     9780002005883  0002005883                     Gilead   
1     9780002261982  0002261987               Spider's Web   
2     9780006163831  0006163831               The One Tree   
3     9780006178736  0006178731             Rage of angels   
4     9780006280897  0006280897             The Four Loves   
...             ...         ...                        ...   
6805  9788185300535  8185300534                  I Am that   
6806  9788185944609  8185944601       Secrets Of The Heart   
6807  9788445074879  8445074873             Fahrenheit 451   
6808  9789027712059  9027712050   The Berlin Phenomenology   
6809  9789042003408  9042003405  'I'm Telling You Stories'   

                                            subtitle  \
0                                                NaN   
1                                            A Novel   
2                                                NaN   
3                                                NaN   
4                                                NaN   
...                                              ...   
6805             Talks with Sri Nisargadatta Maharaj   
6806                                             NaN   
6807                                             NaN   
6808                                             NaN   
6809  Jeanette Winterson and the Politics of Reading   

                                           authors  \
0                               Marilynne Robinson   
1                  Charles Osborne;Agatha Christie   
2                             Stephen R. Donaldson   
3                                   Sidney Sheldon   
4                              Clive Staples Lewis   
...                                            ...   
6805  Sri Nisargadatta Maharaj;Sudhakar S. Dikshit   
6806                                 Khalil Gibran   
6807                                  Ray Bradbury   
6808                 Georg Wilhelm Friedrich Hegel   
6809                        Helena Grice;Tim Woods   

                         categories  \
0                           Fiction   
1     Detective and mystery stories   
2                  American fiction   
3                           Fiction   
4                    Christian life   
...                             ...   
6805                     Philosophy   
6806                      Mysticism   
6807                   Book burning   
6808                        History   
6809             Literary Criticism   

                                              thumbnail  \
0     http://books.google.com/books/content?id=KQZCP...   
1     http://books.google.com/books/content?id=gA5GP...   
2     http://books.google.com/books/content?id=OmQaw...   
3     http://books.google.com/books/content?id=FKo2T...   
4     http://books.google.com/books/content?id=XhQ5X...   
...                                                 ...   
6805  http://books.google.com/books/content?id=Fv_JP...   
6806  http://books.google.com/books/content?id=XcrVp...   
6807                                                NaN   
6808  http://books.google.com/books/content?id=Vy7Sk...   
6809  http://books.google.com/books/content?id=2lVyR...   

                                            description  published_year  \
0     A NOVEL THAT READERS and critics have been eag...          2004.0   
1     A new 'Christie for Christmas' -- a full-lengt...          2000.0   
2     Volume Two of Stephen Donaldson's acclaimed se...          1982.0   
3     A memorable, mesmerizing heroine Jennifer -- b...          1993.0   
4     Lewis' work on the nature of love divides love...          2002.0   
...                                                 ...             ...   
6805  This collection of the timeless teachings of o...          1999.0   
6806                                                NaN          1993.0   
6807                                                NaN          2004.0   
6808  Since the th

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6810 entries, 0 to 6809
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   isbn13          6810 non-null   int64  
 1   isbn10          6810 non-null   object 
 2   title           6810 non-null   object 
 3   subtitle        2381 non-null   object 
 4   authors         6738 non-null   object 
 5   categories      6711 non-null   object 
 6   thumbnail       6481 non-null   object 
 7   description     6548 non-null   object 
 8   published_year  6804 non-null   float64
 9   average_rating  6767 non-null   float64
 10  num_pages       6767 non-null   float64
 11  ratings_count   6767 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 638.6+ KB


In [6]:
df.describe()

isbn13  published_year  average_rating    num_pages  \
count  6.810000e+03     6804.000000     6767.000000  6767.000000   
mean   9.780677e+12     1998.630364        3.933284   348.181026   
std    6.068911e+08       10.484257        0.331352   242.376783   
min    9.780002e+12     1853.000000        0.000000     0.000000   
25%    9.780330e+12     1996.000000        3.770000   208.000000   
50%    9.780553e+12     2002.000000        3.960000   304.000000   
75%    9.780810e+12     2005.000000        4.130000   420.000000   
max    9.789042e+12     2019.000000        5.000000  3342.000000   

       ratings_count  
count   6.767000e+03  
mean    2.106910e+04  
std     1.376207e+05  
min     0.000000e+00  
25%     1.590000e+02  
50%     1.018000e+03  
75%     5.992500e+03  
max     5.629932e+06

In [7]:
df.isnull().sum()

isbn13               0
isbn10               0
title                0
subtitle          4429
authors             72
categories          99
thumbnail          329
description        262
published_year       6
average_rating      43
num_pages           43
ratings_count       43
dtype: int64

- mainly books recommending based on description of the books ,there's 262 missing values in description column.so its gonna filled with empty strings.

In [8]:
df['description']=df['description'].fillna('')

- using TfidfVectorizer to convert text descriptions into a TF-IDF matrix.

In [9]:
from sklearn.feature_extraction.text import TfidfVectorizer
tfidf = TfidfVectorizer(stop_words="english")

tfidf_matrix = tfidf.fit_transform(df['description'])


-  retrieve the feature names (terms) from the TF-IDF matrix.

In [10]:
feature_names = tfidf.get_feature_names_out()

- calculates the cosine similarity between pairs of documents using the TF-IDF matrix obtained from the book descriptions.

In [12]:
from sklearn.metrics.pairwise import cosine_similarity
cosine_sim = cosine_similarity(tfidf_matrix, tfidf_matrix)


- A Series is created with book titles as index and their corresponding indices as values. duplicate book titles are removed, keeping only the last occurrence.







In [13]:
indices = pd.Series(df.index, index=df['title'])

In [14]:
indices = indices[~indices.index.duplicated(keep='last')]

**Function to Recommend Books Based on Similarity Scores**

In [18]:
def recommend_books(book_title, num_recommendations=10):
    book_index = indices[book_title]


    similarity_scores = pd.DataFrame(cosine_sim[book_index], columns=["score"])

    recommended_indices = similarity_scores.sort_values("score", ascending=False)[1:num_recommendations+1].index


    return df['title'].iloc[recommended_indices]

- exapmle usage by crime and punishment book

In [19]:
crime_and_punishment_recommendations = recommend_books("Crime and Punishment")
print(crime_and_punishment_recommendations)


845                  The Brothers Karamazov
4501                             Hard Eight
6129                   Notebooks, 1935-1951
6331       The Man who Watched Trains Go by
1295                            Euripides I
3603                            Taxi Driver
2766            The Cricket in Times Square
3053    Paradise Lost and Paradise Regained
1625                             Hard Eight
948                    A Tale of Two Cities
Name: title, dtype: object
